library/Zend/Form/Form.php

```diff
@@ -497,6 +497,12 @@ public function isValid()
             ));
         }
 
+        foreach ($filter->getInputs() as $input) {
+            if (method_exists($input, 'getName') && !$this->has($input->getName())) {
+                $filter->remove($input->getName());
+            }
+        }
+
         $filter->setData($this->data);
         $filter->setValidationGroup(InputFilterInterface::VALIDATE_ALL);
```

Comment:
This should be changed. Here you iterate through the form elements and if they exist you add them to the validation group.